data_steward/cdr_cleaner/cleaning_rules/aggregate_zip_codes.py

```diff
@@ -0,0 +1,235 @@
+"""
+To further obfuscate participant identity, some generalized zip codes will be aggregated together.
+The PII zip code and state will be transformed to a neighboring zip code/state pair for 
+those zip codes with low population density.
+It is expected that this lookup table will be static and will remain unchanged. 
+It is based on US population, and not on participant address metrics.
+
+Original Issues: DC-1379
+"""
+
+# Python imports
+import logging
+
+# Project imports
+import constants.bq_utils as bq_consts
+from cdr_cleaner.cleaning_rules.base_cleaning_rule import BaseCleaningRule
+from cdr_cleaner.cleaning_rules.generalize_zip_codes import GeneralizeZipCodes
+from cdr_cleaner.cleaning_rules.deid.string_fields_suppression import StringFieldsSuppression
+from constants.cdr_cleaner import clean_cdr as cdr_consts
+from common import JINJA_ENV, OBSERVATION, PIPELINE_TABLES
+from utils import pipeline_logging
+
+LOGGER = logging.getLogger(__name__)
+PII_STATE_VOCAB = 'pii_state_vocab'
+ZIP_CODE_AGGREGATION_MAP = 'zip_code_aggregation_map'
+ZIP_CODES_AND_STATES_TO_MODIFY = '_zip_codes_and_states_to_modify'
+
+FIND_ZIP_CODES_AND_STATES_QUERY = JINJA_ENV.from_string("""
```

Comment:
You can add in a separate query to sandbox the original records, below is an example, that I think should work based off your original query.
```
SANDBOX_QUERY = JINJA_ENV.from_string("""
SELECT
    zip_code_obs.*
FROM `{{project_id}}.{{dataset_id}}.{{obs_table}}` zip_code_obs
JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{zip_code_aggregation_map}}` map
    ON map.zip_code_3 = zip_code_obs.value_as_string
JOIN `{{project_id}}.{{pipeline_tables_dataset}}.{{pii_state_vocab}}` state_vocab
    ON state_vocab.postal_code = map.state
LEFT JOIN `{{project_id}}.{{dataset_id}}.{{obs_table}}` state_obs
    ON state_obs.person_id = zip_code_obs.person_id
        AND state_obs.observation_source_concept_id = 1585249
        AND state_vocab.concept_id = state_obs.value_source_concept_id
WHERE zip_code_obs.observation_source_concept_id = 1585250
"""
```

Reply:
This makes sense for sandboxing the zip code records. Should the state records be sandboxed as well, maybe in a UNION?

Reply:
Whoops my bad, yeah I think that should work!

Playing around I think the following could work:

```
SELECT
  zip_code_obs.*
FROM
  `{{project_id}}.{{dataset_id}}.{{obs_table}}` zip_code_obs
WHERE
  zip_code_obs.observation_source_concept_id = 1585250
  AND zip_code_obs.value_as_string IN(
  SELECT
    zip_code_3
  FROM
    `{{project_id}}.{{pipeline_tables_dataset}}.{{zip_code_aggregation_map}}` )
UNION ALL
SELECT
  state_obs.*
FROM
  `{{project_id}}.{{dataset_id}}.{{obs_table}}` state_obs
WHERE
  state_obs.observation_source_concept_id = 1585249
  AND state_obs.person_id IN (
  SELECT
    person_id
  FROM
    `{{project_id}}.{{dataset_id}}.{{obs_table}}`
  WHERE
    observation_source_concept_id = 1585250
    AND value_as_string IN(
    SELECT
      zip_code_3
    FROM
      `{{project_id}}.{{pipeline_tables_dataset}}.{{zip_code_aggregation_map}}`))
```

I feel like this is a little redundant and I just quickly wrote up, so if you have a more efficient and elegant way of doing it (which I feel like from your queries in this ticket, you do :) ), feel free to adjust!

Reply:
Thanks :) . Yep, this is something like what I had in mind.